add ability to run tests and linting in new projects

aditya-padhi-kbl opened the pull request:
Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

Resolves: #92 


Code changes:-

Copy over entiretest directory from reaction core
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-d2e63c5855748bce4cbde6c971b64b813baea8c7c6e0a62a2926c6a9afaeda95R1
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-e2e7861e8c19ccb501921a47ca655b1fc48e7464320f86aafae63ed3239f8ceaR109

Move checkNodeVersion.csj from "utils" to "src"
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-e2e7861e8c19ccb501921a47ca655b1fc48e7464320f86aafae63ed3239f8ceaR101

Add tests scripts to package.json
** it already exists **

Copy over jest config
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-e2e7861e8c19ccb501921a47ca655b1fc48e7464320f86aafae63ed3239f8ceaR111

Copy over babel config
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-e2e7861e8c19ccb501921a47ca655b1fc48e7464320f86aafae63ed3239f8ceaR113

Needs to have Mongo running. The Jest setup needs to set the MONGO_URL to MONGO_URL=mongodb://localhost:27017/test
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-68b75959b4ea738c8e466b862b12daea4e18d213e2c1f15884f68977c6b4fc7e

aditya-padhi-kbl committed:
fixes for issue no. 92. changes in create-project-api and added utils

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

brent-hoover commented on commands/create-project-api.js:
```diff
@@ -70,7 +73,7 @@ async function updatePackageJson(packageJson, projectName) {
 function updateEnv(envData) {
   const env = parse(envData);
-  env.MONGO_URL = "mongodb://localhost:27017/reaction";
+  env.MONGO_URL = "mongodb://localhost:27017/test";
```
This change the permanent link and cause production data and test data to comingle. This change needs to just change the the MONGO for the test runner only

aditya-padhi-kbl replied:
In the jest config, there is no reference for MONGO . Should I refer to some other location ?

aditya-padhi-kbl replied:
Hi @zenweasel I need some help regarding setting up of MONGO URL for test runner.  Can you please provide some direction

brent-hoover replied:
Jest has a way to set ENV VARS just for the duration of the test run that is not specific to this one. You just need to apply that method, I don't know it off the top of my head.

aditya-padhi-kbl replied:
@zenweasel I am thinking we have to update Mongo test URL here:- https://github.com/reactioncommerce/reaction/blob/trunk/apps/reaction/tests/util/setupJestTests.js

As per the jest documentation this is the way to update environment variables

brent-hoover replied:
You need to update after you copy it into the project, not touch the original project files

aditya-padhi-kbl replied:
@zenweasel do you mean updating the setUpJestTests.js file of the project created using cli ?

aditya-padhi-kbl replied:
Is it fine if I completely update the file with :-

`
process.env = Object.assign(process.env, {
  MAIL_URL: "smtp://user:pass@email-smtp.us-west-2.amazonaws.com:465",
  REACTION_LOG_LEVEL: "ERROR",
  REACTION_WORKERS_ENABLED: false,
MONGO_URL: "mongodb://localhost:27017/test"
});

process.on("unhandledRejection", (err) => {
  throw err;
});
`

The only downside of it is that if the source file content changes, we may have compatibility issues.
Other solution is in the .env file we can use a new property like:-
`
  MONGO_TEST_URL: "mongodb://localhost:27017/test"
`

brent-hoover replied:
I would suggest just adding the MONGO_URL entry after copying the file over

aditya-padhi-kbl replied:
The challenge is around adding it to a javascript file [here](https://github.com/reactioncommerce/reaction/blob/trunk/apps/reaction/tests/util/setupJestTests.js).  If we would have imported a json file & used it internally, it would have been very easy.

brent-hoover replied:
Yeah, probably the least worst solution is to add the file with the change to the project template and just use that.

aditya-padhi-kbl replied:
instead should I create a pull request in **reaction** to import a json [here](https://github.com/reactioncommerce/reaction/blob/trunk/apps/reaction/tests/util/setupJestTests.js#L5) & use it like:-


```
import { createRequire } from "module";
const require = createRequire(import.meta.url);
const jestConfigs = require("./jestConfigs.json");
process.env = Object.assign(process.env, {
  ...jestConfigs
});

process.on("unhandledRejection", (err) => {
  throw err;
});
```


We can use it later in our repository & simply update the `jestConfigs.json`
It will help to reduce the maintenance overhead in cli

aditya-padhi-kbl replied:
I updated the template here:-
https://github.com/reactioncommerce/cli/compare/trunk...aditya-padhi-kbl:cli:trunk?expand=1#diff-68b75959b4ea738c8e466b862b12daea4e18d213e2c1f15884f68977c6b4fc7e

I feel if we merge https://github.com/reactioncommerce/reaction/pull/6568 then we don't have to maintain test template on our side

aditya-padhi-kbl replied:
@zenweasel 
Can you please review this [PR](https://github.com/reactioncommerce/reaction/pull/6568) & merge if the changes look good.  Once done I will go ahead & make changes to MONGO_URL integration

aditya-padhi-kbl replied:
Hi @zenweasel ,

Please review the changes. Mongo_url integration & other changes are done

brent-hoover commented on utils/copyTests.js:
```diff
@@ -0,0 +1,13 @@
+import { fileURLToPath } from "url";
```
This code seems to be copying over the tests for the CLI when it actually needs to be copying over the tests from the API project.

aditya-padhi-kbl replied:
Thank-you for the details.  In that case should I hardcode the urls & fetch it using getFilesFromRepo ?

aditya-padhi-kbl replied:
Done

brent-hoover replied:
To clarify further you need to copy all the file from [here](https://github.com/reactioncommerce/reaction/tree/trunk/apps/reaction/tests)

aditya-padhi-kbl replied:
@zenweasel I am thinking we have to update Mongo URL here:- https://github.com/reactioncommerce/reaction/blob/trunk/apps/reaction/tests/util/setupJestTests.js

brent-hoover commented on commands/create-project-api.js:
```diff
@@ -5,6 +5,8 @@ import getFilesFromRepo from "../utils/getFilesFromRepo.js";
 import Logger from "../utils/logger.js";
 import getFileFromCore, { reactionRoot } from "../utils/getFileFromCore.js";
+import copyTests from "../utils/copyTests.js";
+import { babelConfigurationFileName, jestFileName } from "./constants.js";
```
The convention is if these are actually constants to use ALL_UPPER_CASE. However since we don't do this anywhere else and the name of these files won't change I would suggest just defining them inline.

aditya-padhi-kbl replied:
Done

brent-hoover reviewed (changes requested):
Thanks for this! A few changes here

aditya-padhi-kbl committed:
-s

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl committed:
copied test cases from reaction

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl committed:
added testTemplates. Used it for project jest setup

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl committed:
fix: changes to copy tests folder. checkNodeVersion to src.

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl committed:
Merge branch 'reactioncommerce:trunk' into trunk

aditya-padhi-kbl committed:
Merge branch 'reactioncommerce:trunk' into trunk

aditya-padhi-kbl commented:
@zenweasel , @delagroove Please review the PR.  Please suggest if any changes are needed

brent-hoover commented on utils/copyTests.js:
```diff
@@ -0,0 +1,33 @@
+  return true;
+};
+
+export default copyTests;
```
Can you follow the export style of the rest of the app?

aditya-padhi-kbl replied:
sure

aditya-padhi-kbl replied:
@zenweasel Somehow this pull request is closed.  I have absolutely no idea how this happened

aditya-padhi-kbl replied:
Just got to know that after force-push the PR got resolved.  I don't have the rights to reopen the closed PR.  Can the admins please help to reopen the closed PR:- (https://github.com/reactioncommerce/cli/pull/133)

aditya-padhi-kbl replied:
@zenweasel This PR is open now. :). Thank-you for helping :)
https://github.com/aditya-padhi-kbl/cli_forked/blob/trunk/utils/copyTests.js#L16

brent-hoover commented:
@aditya-padhi-kbl Just heads up that I won't have time to review and test until tomorrow

brent-hoover commented:
@aditya-padhi-kbl When I try to create a project and run `npm run test` I get 100's of these messages:


``` 
 FAIL  tests/integration/api/queries/shippingRates/flatRateFulfillmentMethod.test.js
  ● Test suite failed to run

    /home/brent/Projects/merchstack/cli-test-projects/oct-20-2022/myserver/tests/integration/api/queries/shippingRates/flatRateFulfillmentMethod.test.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
                                                                                            ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      at ScriptTransformer._transformAndBuildScript (../../../../../../../usr/lib/node_modules/jest/node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
```
Is `npm run test` working for you when you create a project?

brent-hoover commented:
@aditya-padhi-kbl Ok, I think I see the issue. Since I originally created this issue we have restructured the main Reaction repo and the babel, eslint and jest dependencies all live at the root level, not in the /apps/reaction/package.json. 

I added these locally to the project I created with your fork and tests ran fine.

Basically I added all the packages from the devDependencies [here](https://github.com/reactioncommerce/reaction/blob/trunk/package.json) and removed `    "graphql-schema-linter" (because we aren't supporting it here)

We also need to copy over the eslint config file from the [here](https://github.com/reactioncommerce/reaction/blob/trunk/.eslintrc.cjs) and that should make `npm run lint` work as well

aditya-padhi-kbl commented:
> @aditya-padhi-kbl Ok, I think I see the issue. Since I originally created this issue we have restructured the main Reaction repo and the babel, eslint and jest dependencies all live at the root level, not in the /apps/reaction/package.json.
> 
> I added these locally to the project I created with your fork and tests ran fine.
> 
> Basically I added all the packages from the devDependencies [here](https://github.com/reactioncommerce/reaction/blob/trunk/package.json) and removed ` "graphql-schema-linter" (because we aren't supporting it here)
> 
> We also need to copy over the eslint config file from the [here](https://github.com/reactioncommerce/reaction/blob/trunk/.eslintrc.cjs) and that should make `npm run lint` work as well

@zenweasel 
Oh got it.
In that case should we just copy the eslint config file or copy the devDependencies from [here](https://github.com/reactioncommerce/reaction/blob/trunk/package.json) as well ?

brent-hoover commented:
We need to copy the devDependencies in order for tests and linting to work

aditya-padhi-kbl commented:
> 

Ok.  In that case I think we shouldn't  copy package-lock.json because the details present in the lock file & in package.json are different. What do you suggest ?

aditya-padhi-kbl commented:
> We need to copy the devDependencies in order for tests and linting to work

I am getting this result.  Somewhere the dependencies are conflictiing
```
apadhi@macos-C02FW1H5MD6N test_reaction_api % npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: test_reaction_api@1.0.0
npm ERR! Found: eslint-plugin-import@2.26.0
npm ERR! node_modules/eslint-plugin-import
npm ERR!   dev eslint-plugin-import@"^2.26.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint-plugin-import@"~2.25.4" from @reactioncommerce/eslint-config@2.2.2
npm ERR! node_modules/@reactioncommerce/eslint-config
npm ERR!   dev @reactioncommerce/eslint-config@"~2.2.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/apadhi/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/apadhi/.npm/_logs/2022-10-21T06_50_04_721Z-debug-0.log
apadhi@macos-C02FW1H5MD6N test_reaction_api % 
```

brent-hoover commented:
Here is what I have in `devDependencies` and it's working for me:
```
    "@reactioncommerce/data-factory": "~1.0.1",
    "@reactioncommerce/eslint-config": "~2.2.0",
    "apollo-link-http": "~1.5.16",
    "apollo-server-testing": "~2.9.6",
    "faker": "~4.1.0",
    "graphql-tools": "~4.0.5",
    "nock": "~11.4.0",
    "node-fetch": "~2.6.0",
    "pinst": "^2.1.4",
    "wait-for-expect": "~3.0.0",
    "@babel/cli": "^7.18.10",
    "@babel/core": "^7.18.10",
    "@babel/eslint-parser": "^7.18.9",
    "@babel/plugin-transform-modules-commonjs": "^7.18.6",
    "@babel/preset-env": "^7.19.0",
    "@commitlint/config-conventional": "^17.0.3",
    "@reactioncommerce/babel-remove-es-create-require": "^1.0.0",
    "babel-jest": "^29.0.1",
    "babel-plugin-add-module-exports": "^1.0.4",
    "babel-plugin-rewire-exports": "^2.3.0",
    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
    "babel-plugin-transform-import-meta": "^2.2.0",
    "eslint": "^7.32.0",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jest": "^26.8.3",
    "eslint-plugin-jsx-a11y": "^6.6.1",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-promise": "^6.0.1",
    "eslint-plugin-react": "^7.31.1",
    "eslint-plugin-react-hooks": "^4.6.0",
    "husky": "^5.0.8",
    "is-ci": "^2.0.0",
    "is-docker": "^2.1.1",
    "jest": "^25.5.4",
    "nodemon": "~1.19.2"
```

aditya-padhi-kbl commented:
> > 
> 
> Ok. In that case I think we shouldn't copy package-lock.json because the details present in the lock file & in package.json are different. What do you suggest ?

@zenweasel Please confirm if we need package-lock to be copied as it will not contain the references of devDependencies

brent-hoover commented:
We should not copy the package-lock. They should generate their own

aditya-padhi-kbl committed:
fix: synced devdependencies and copied tests and eslintrc and some minor cleanup

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl commented:
> We should not copy the package-lock. They should generate their own



> We should not copy the package-lock. They should generate their own

@zenweasel 
I resolved the issues.  Test cases are running now. Please review

aditya-padhi-kbl commented:
@zenweasel Please let me know if the changes look good.

brent-hoover commented:
I am seeing the same error you were seeing. Trying to figure out the best way to resolve this.
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: myserver@1.0.0
npm ERR! Found: eslint-plugin-import@2.26.0
npm ERR! node_modules/eslint-plugin-import
npm ERR!   dev eslint-plugin-import@"^2.26.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint-plugin-import@"~2.25.4" from @reactioncommerce/eslint-config@2.2.2
npm ERR! node_modules/@reactioncommerce/eslint-config
npm ERR!   dev @reactioncommerce/eslint-config@"~2.2.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/brent/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/brent/.npm/_logs/2022-10-24T06_47_37_661Z-debug-0.log

```

aditya-padhi-kbl commented:
> eslint-plugin-import@"^2.26.0" from the root project
> npm ERR! 
> npm ERR! Could not resolve dependency:
> npm ERR! peer eslint-plugin-import@"~2.25.4" from @reactioncommerce/eslint-config@2.2.2
> npm ERR! node_modules/@reactioncommerce/eslint-config
> npm ERR!   dev @reactioncommerce/eslint-config@"~2.2.0" from the root project
> npm ERR! 
> npm ERR! Fix the upstream dependency conflict, or retry
> npm ERR! this command with --force, or --legacy-peer-deps
> npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
> npm ERR! 
> npm ERR! See /home/brent/.npm/eresolve-report.txt for a full report.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /home/brent/.npm/_logs/2022-10-24T06_47_37_661Z-debug-0.log

I think you will not get this issue when you pull the latest changes from my fork.
Here is the same screen recording:-

brent-hoover commented:
@aditya-padhi-kbl I did actually pull your latest changes and create a new project

brent-hoover commented:
The suggestion from @vanpho93 is to copy in the `.npmrc` and add a new line for install legacy peer deps. Worked for me locally

<img width="427" alt="image (1)" src="https://user-images.githubusercontent.com/72819/197469734-b1887437-c877-4b07-96d4-187568496002.png">

aditya-padhi-kbl commented:
> The suggestion from @vanpho93 is to copy in the `.npmrc` and add a new line for install legacy peer deps. Worked for me locally
> 
> <img alt="image (1)" width="427" src="https://user-images.githubusercontent.com/72819/197469734-b1887437-c877-4b07-96d4-187568496002.png">

Sure let me try this. Somehow I couldn't reproduce it locally. Here is a sample recording:- https://drive.google.com/file/d/1C5e-nXbu607JGWCeORU9UW6cpMH-OPyS/view?usp=sharing

aditya-padhi-kbl committed:
fix: added and updated npmrc from reactionroot

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl commented:
> > The suggestion from @vanpho93 is to copy in the `.npmrc` and add a new line for install legacy peer deps. Worked for me locally
> > <img alt="image (1)" width="427" src="https://user-images.githubusercontent.com/72819/197469734-b1887437-c877-4b07-96d4-187568496002.png">
> 
> Sure let me try this. Somehow I couldn't reproduce it locally. Here is a sample recording:- https://drive.google.com/file/d/1C5e-nXbu607JGWCeORU9UW6cpMH-OPyS/view?usp=sharing

@zenweasel I copied npmrc & I think it is working fine now.
https://github.com/reactioncommerce/cli/pull/133/files#diff-e2e7861e8c19ccb501921a47ca655b1fc48e7464320f86aafae63ed3239f8ceaR90

Please review

aditya-padhi-kbl committed:
Merge branch 'reactioncommerce:trunk' into trunk

brent-hoover reviewed (changes requested):
@aditya-padhi-kbl Tested and works great. Great work and thanks for sticking with this. Just one little style fix and then were ready to release.

aditya-padhi-kbl committed:
fix: fixed jsdoc issues in create-project. review fix function export style

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

aditya-padhi-kbl committed:
Merge branch 'trunk' of github.com:aditya-padhi-kbl/cli_forked into trunk. resolved code conflicts and lint errors

Signed-off-by: aditya-padhi-kbl <aditya.padhi@outlook.com>

rc-publisher commented:
:tada: This PR is included in version 1.5.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/reaction-cli/v/1.5.1)
- [GitHub release](https://github.com/reactioncommerce/cli/releases/tag/v1.5.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
